File: lib/features/sessions/domain/active_session.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'active_session.freezed.dart';

@freezed
abstract class ActiveSession with _$ActiveSession {
  const factory ActiveSession({
    required int pursuitId,
    required DateTime startedAt,
    @Default(Duration.zero) Duration pausedTotal,
    DateTime? pauseStartedAt,
  }) = _ActiveSession;

  const ActiveSession._();

  bool get isPaused => pauseStartedAt != null;

  Duration elapsedAt(DateTime now) {
    final activePauseSoFar = pauseStartedAt == null
        ? Duration.zero
        : now.difference(pauseStartedAt!);
    final raw = now.difference(startedAt) - pausedTotal - activePauseSoFar;
    return raw.isNegative ? Duration.zero : raw;
  }

  // The instant this session should be recorded as ending so cumulative
  // covered time lands exactly on [target], never banking a background
  // overshoot. [priorCounted] is the counted duration from *other* completed
  // sessions for the pursuit. Returns null when no clamp applies — paused, the
  // target was already met before this session, or the crossing isn't in the
  // past at [now] (caller should just use [now] then).
  DateTime? completionEndAt({
    required Duration priorCounted,
    required Duration target,
    required DateTime now,
  }) {
    if (isPaused) return null;
    final remaining = target - priorCounted;
    if (remaining <= Duration.zero) return null;
    final crossing = startedAt.add(pausedTotal).add(remaining);
    return crossing.isBefore(now) ? crossing : null;
  }
}

File: Sources/Models/ActiveSession.swift
import Foundation

// The single running/paused timer (pure domain value). Elapsed is *always*
// wall-clock timestamp math — never a tick counter — so it survives
// backgrounding, kill, DST, and reboot with no special handling.
// Ported from sessions/domain/active_session.dart.
struct ActiveSession: Equatable {
  var pursuitId: Int
  var startedAt: Date
  /// Total accumulated pause time from completed pause intervals, in seconds.
  var pausedTotal: TimeInterval = 0
  /// Set while a pause is in progress; nil when running.
  var pauseStartedAt: Date?

  var isPaused: Bool { pauseStartedAt != nil }

  /// elapsed = now - startedAt - pausedTotal - (now - pauseStartedAt if paused),
  /// clamped to ≥ 0 for clock-skew defensiveness.
  func elapsed(at now: Date) -> TimeInterval {
    let activePauseSoFar = pauseStartedAt.map { now.timeIntervalSince($0) } ?? 0
    let raw = now.timeIntervalSince(startedAt) - pausedTotal - activePauseSoFar
    return max(0, raw)
  }

  /// The instant this session should be recorded as ending so cumulative covered
  /// time lands exactly on `target`, never banking a background overshoot.
  /// `priorCounted` is the counted duration from *other* completed sessions for
  /// the pursuit. Returns nil when no clamp applies — paused, the target was
  /// already met before this session, or the crossing isn't yet in the past at
  /// `now` (caller should just use `now` then).
  func completionEndAt(priorCounted: TimeInterval, target: TimeInterval, now: Date) -> Date? {
    if isPaused { return nil }
    let remaining = target - priorCounted
    if remaining <= 0 { return nil }
    let crossing = startedAt.addingTimeInterval(pausedTotal + remaining)
    return crossing < now ? crossing : nil
  }
}
